dcr/wallet.go

```diff
@@ -2,38 +2,93 @@ package dcr
 
 import (
 	"context"
+	"encoding/hex"
 	"fmt"
 	"path/filepath"
 	"sync"
+	"time"
 
+	"decred.org/dcrdex/client/mnemonic"
 	"decred.org/dcrwallet/v4/spv"
 	"decred.org/dcrwallet/v4/wallet"
 	"github.com/decred/dcrd/chaincfg/v3"
-	"github.com/decred/libwallet/asset"
 	"github.com/decred/slog"
 )
 
 type mainWallet = wallet.Wallet
 
 type Wallet struct {
-	*asset.WalletBase
 	dir         string
 	dbDriver    string
 	chainParams *chaincfg.Params
 	log         slog.Logger
 
-	db wallet.DB
+	metaData *walletData
+	db       wallet.DB
 	*mainWallet
 
 	syncerMtx sync.RWMutex
 	syncer    *spv.Syncer
+	*syncHelper
 }
 
 // MainWallet returns the main dcr wallet with the core wallet functionalities.
 func (w *Wallet) MainWallet() *wallet.Wallet {
 	return w.mainWallet
 }
 
+// DecryptSeed decrypts the encrypted wallet seed using the provided passphrase
+// and returns the mnemonic.
+func (w *Wallet) DecryptSeed(passphrase []byte) (string, error) {
+	w.metaData.seedMtx.Lock()
+	defer w.metaData.seedMtx.Unlock()
+
+	if w.metaData.EncryptedSeedHex == "" {
+		return "", fmt.Errorf("seed has been verified")
+	}
+
+	encryptedSeed, err := hex.DecodeString(w.metaData.EncryptedSeedHex)
+	if err != nil {
+		return "", fmt.Errorf("unable to decode encrypted hex seed: %v", err)
+	}
+
+	seed, err := DecryptData(encryptedSeed, passphrase)
+	if err != nil {
+		return "", err
+	}
+	return mnemonic.GenerateMnemonic(seed, time.Unix(w.metaData.Birthday, 0))
+}
+
+func (w *Wallet) ReEncryptSeed(oldPass, newPass []byte) error {
```

Comment:
Are we not using this? Should be used when changing password I guess? Is this a huge mistake?

Reply:
#27